Initial commit data dictionary

Ashleigh234 opened the pull request:
Data Dictionary documentation, extracting and verification tools. I wasn't sure exactly where to put the documentation and tools so have created a new folder to put them all in. 

There are the comment_extraction and markdown_check scripts as well as a markdown_auto which can automatically run the checks and convert the markdown files to the associated pdf, readthedocs markdown version and created the sql comments.  There is also a latex template which adjusts the default pdf generation.

imincik commented on tools/comment_extraction.py:
```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/python
+
+'''
+Python script to extract notes from tables within markdown document to extract as comments
```
* small typo here - missing space in `PostgreSQLcomment`

imincik commented on tools/markdown_auto.sh:
```diff
@@ -0,0 +1,40 @@
+# Script to automatically check markdown file and then:
+#   - pdf from markdown file, using latex variables \newline \newpage.
+#   - commonmark file for use in readthedocs, with table converted to html.
+#   - postgreSQLcomments sql file, extraction of table comments from markdown.
```
* the same typo as above

strk commented:
I'll review next week, meanwhile I can suggest to put
the documentation files under a doc/ directory.

strk commented:
Another comment: I just noticed you pushed your work directly on the main repository, rather than in a branch of our fork. I think this should be discouraged, in general.

strk commented:
I don't think the example outputs should be in the repository, and not even sure the example input should be. Also what is the png diagram for ? 

All checks should be run as part of `make check`. Comments generation should be run as part of `make` and produce another file to be installed by `make install` and loaded by `linz-lds-bde-schema-loader`; doing so should also be testing quality of comments which would end up being loaded by the tester (to be checked that this is the case). Tests could be added to ensure comments do exist.

Some instructions for the input markdown editor should be provided.

Ashleigh234 commented:
I have updated the locations, by splitting documentation into /doc, the png models used in the pdf spit into /models. 
Updated Makefile to run checks through make check, comment extraction and update pdf files through make. 
Added a file inside /doc giving details for editing the markdown files.
Updated the markdown data dictionaries to be full conversions of the original pdf data dictionaries, rather than the previous example versions.
Updated to wrap lines to 80 characters where it was possible.

palmerj commented on tools/comment_extraction.py:
```diff
@@ -0,0 +1,142 @@
+Python script to extract notes from tables within markdown document to extract
+as comments to PostgreSQL comment.
+
+Currently just converts table/ attribute name to lowercase and uses that as
```
This assumption is not correct. You will need to use a mapping file between actual database table names and the names used in the data dictionary documentation. I think a good test is creating the LDS and BDE schema tables in a database and then trying to apply the comments files

Ashleigh234 replied:
I had thought that was the case needing to be mapped to the actual stored names, just had converted the documentation names directly as an initial test to generate the comments before I knew what the mappings should be. I have been able to map all the names from the documentation to the actual database table names except for Ordinate Adjustment table which I couldn't find the mapping for.

palmerj replied:
Oh yes Ordinate Adjustment is a new table this year and won't be in that old document. the source table is bde.crs_ordinate_adj

palmerj commented:
Great work so far @Ashleigh234. Great to have the readme file on the structure and tips for maintaining the documentation sources. Couple of quick things I see without doing any code review which I will leave to @strk:

1. Why are the Rich Text Document version of the markdown stored in the repository? 
2. Can you please add cell boarders to the table styles in the output pdf? Only having horizontal lines for the header row and bottom of the table makes alignment of rows and columns in large tables hard

strk commented on Makefile:
```diff
@@ -20,7 +20,8 @@ SQLSCRIPTS = \
   sql/99-patches.sql \
-  sql/versioning/01-version_tables.sql
+  sql/versioning/01-version_tables.sql \
+  sql/lds_comments.sql \
```
As sql/lds_comments.sql is a generated file it should be added to SCRIPTS_built, rather than to SCRIPTS.
Then adding a rule to build it should make the rest automatic.

strk commented on Makefile:
```diff
@@ -33,10 +34,22 @@ EXTRA_CLEAN = \
+	tools/markdown_auto.sh \
+	  doc/lds-full-landonline-data-dictionary-and-models.md
+	tools/markdown_auto.sh \
+	  doc/property-and-ownership-simplified-tables-data-dictionary.md
```
it isn't clear what the output of `tools/markdown_auto.sh` is here.
You should tipically form Makefile rules like:

```
 output: inputs
    command to generate output from inputs
```

strk commented on Makefile:
```diff
@@ -60,6 +75,10 @@ uninstall:
+	tools/markdown_check.py \
+	  doc/lds-full-landonline-data-dictionary-and-models.md
+	tools/markdown_check.py \
+	  doc/property-and-ownership-simplified-tables-data-dictionary.md
```
Please put the documentation checks in their own rule (check-docs?) and make this new rule a dependency of the "check test" rule. Just to keep things more readable.

strk commented on doc/README.md:
```diff
@@ -0,0 +1,107 @@
+    title: Dictionary Title 
+    subtitle: Dictionary Subtitle
+    date: Date \newline Version
+    ---
```
This document doesn't say anything about which structure is expected by the conversion tools in order to be able to generate table COMMENTS. I'd expect to find in there detailed documentation about that: how do I document a (database) table. The section about "tables" above seems to be instead generally about how to render tables in markdown (nothing specific to Database Tables)

palmerj commented:
> I spent quite a bit of time trying to edit the latex template to get cell borders on the pdf document, but can't seem to get any additional borders other than the current horizontal lines. I found that other people have similar issues online trying to add vertical borders in particular using latex and pandoc. I did find that shading could be added to alternate rows to help distinguish between them as well as spacing between cells can be increased.

As per  https://stackoverflow.com/questions/27300329/table-with-borders-in-html-and-latex-output-from-markdown-source-with-pandoc does adding CSS to your output HTML tables before converting to PDF work?

Ashleigh234 commented:
The mapping between actual database table names and the names used in the data dictionary documentation have all added into the creation of the comments.

I have updated the Makefile and documentation for editing the markdown files. 

I have also managed to get borders on the tables by using a different pandoc pdf engine. Using wkhtmltopdf instead of the default pdflatex, which meant then a css stylesheet could be used instead of the previous latex template. This caused a few problems with newlines, the way images displayed and are referenced  and page breaks when converting to pdf. I have been able to fix these issues and the output pdf look very similar to the original ones.

palmerj commented:
Thanks @Ashleigh234. Table style and table mappings look great - well done!

In regards to the page layout in the PDF, I think there pages which only have tiles on them, due to large preceding tables or images. I think this needs a few tweaks. 

We need to think about storing the PDF and https://readthedocs.org/ generated files from the make process in this repo. Also some more planning about how we will publish to https://readthedocs.org/ should also be discussed and documented. 

Beyond this I think @strk needs to review the new code and @vlindsay needs to look at the output PDF to see what other changes might be required :)

Thanks heaps

strk commented:
I think _no_ generated file should be in the git repository. They can be included in the debian package (ideally a separate package: linz-lds-bde-schema-docs).

I would put "models" and "tools" under doc/ directory as they are specific to docs

It looks like I already placed my review, but no action was taken. Is @Ashleigh234 still working with us or shall I take over the branch ?

palmerj commented:
> It looks like I already placed my review, but no action was taken. Is @Ashleigh234 still working with us or shall I take over the branch ?

She is on leave this week, but returning. I understand her period of work with us has been extended by another 4-6 weeks.

imincik commented:
> I think no generated file should be in the git repository. They can be included in the debian package (ideally a separate package: linz-lds-bde-schema-docs).

That makes a sense for me.

Ashleigh234 commented:
I have updated the pdf output to no longer have the empty pages that were being created, there is still a slight issue if tables are longer then the page length, in which case they are split through parts of cells. I have increased the padding in the cells, so that in the cases that this happens usually no text is split or unreadable. 

I have also updated the makefile, from the previous review. I had thought I had done this last time, but hadn't included the updated makefile.

I have also moved everything under the /doc directory, although I have left the generated files included for now as I'm not completely sure what you are wanting done with them.

strk commented on Makefile:
```diff
@@ -20,23 +20,32 @@ SQLSCRIPTS = \
 
 SCRIPTS_built = \
     scripts/linz-lds-bde-schema-load \
+    sql/07-lds_version.sql \
```
Good catch, but I think you also want to remove 07-lds_version.sql from `SQLSCRIPTS` variable.

Ashleigh234 replied:
I don't recall changing this originally, but the Travis build failed when I removed it from the `SQLSCRIPTS` variable

Ashleigh234 replied:
The place where it is failing now is different to what it was, but is due to `pandoc: command not found` do I need to include add pandoc package in the travis.yml file somewhere?

strk replied:
Yes, in the `before_install` step, you can append to the `sudo apt-get install -y pgtap pgxnclient` line

strk replied:
By the way, please *before* having travis install `pandoc` make sure that the script trying to use pandoc actually *fails* (returns an error code, non zero) because right now it is exiting with a SUCCESS error code which is wrong.

Ashleigh234 replied:
I have added `pandoc` to the `sudo apt-get install -y pgtap pgxnclient` line, but the version it is using is quite old and doesn't contain the `wkhtmltopdf` pdf reader options that the newer version has. I tried to add in `pandoc=2.1.1-1` which works when I run it but in travis is says there is no version `pandoc 2.1.1-1`. Is there another way to change the version it uses?

strk commented on Makefile:
```diff
@@ -20,23 +20,32 @@ SQLSCRIPTS = \
+    $(END)
+    
+TOOLS = \
+    doc/tools/markdown_auto.sh \
```
could you find a better name for that script ? What does it do ? "markdown_auto" doesn't tell me anything.

Ashleigh234 replied:
I have removed the markdown_auto script as most of what it was doing was able to be placed directly into the Makefile itself.

strk commented on Makefile:
```diff
@@ -48,7 +57,14 @@ scripts/linz-lds-bde-schema-load: scripts/linz-lds-bde-schema-load.in Makefile
+	chmod +x doc/tools/comment_extraction.py
+	rm -f sql/lds_comments.sql
+	$(TOOLS) doc/lds-full-landonline-data-dictionary-and-models.md
+	$(TOOLS) doc/property-and-ownership-simplified-tables-data-dictionary.md
```
Could you directly invoke python rather than adding execute bit to scripts ?
Can you please use another variable name for TOOLS ? It looks like being a specific executable, not a generic (set of) tools.
The Makefile rule should ideally be a target, something that is _produced_ by the recipe below it, so rather than calling it `create-docs`, could you give it the name of the output ? Use 2 rules instead of one, if you have 2 outputs (as it looks to be the case). And list all inputs in the dependency part (after the comma). The goal is to have those outputs be regenerated only when needed (if any of the dependencies were updated more recently than the output)

Ashleigh234 replied:
I updated the Makefile so that the rules should all now only be creating single outputs which are named the outputs and are produced by the recipe below, rather than what was happening before.

strk commented on Makefile:
```diff
@@ -48,7 +57,14 @@ scripts/linz-lds-bde-schema-load: scripts/linz-lds-bde-schema-load.in Makefile
+	$(TOOLS) doc/lds-full-landonline-data-dictionary-and-models.md
+	$(TOOLS) doc/property-and-ownership-simplified-tables-data-dictionary.md
+
+install: create-docs $(SQLSCRIPTS) $(SCRIPTS_built)
```
Rather than listing `create-docs` as a dependency here, you would list the targets (generated files) that are needed for the install (ie: `$(DOCS_built) ?) -- btw, comments are already part of `SCRIPTS_built` so you would only need to put the other outputs in a `DOCS_built` variable.

strk commented on Makefile:
```diff
@@ -59,12 +75,18 @@ install: $(SQLSCRIPTS) $(SCRIPTS_built)
+check-docs: $(CHECKS)
+	chmod +x $(CHECKS)
+	$(CHECKS) doc/lds-full-landonline-data-dictionary-and-models.md
+	$(CHECKS) doc/property-and-ownership-simplified-tables-data-dictionary.md
```
Again please use sensible variable names. `CHECKS` here looks like being a single script, give the variable (as well as the script) a meaningful name

strk commented on Makefile:
```diff
@@ -48,7 +57,38 @@ scripts/linz-lds-bde-schema-load: scripts/linz-lds-bde-schema-load.in Makefile
+doc/lds-full-landonline-data-dictionary-and-models.pdf: \
+    doc/tools/markdown-to-pdf-conversion.sh \
+    doc/lds-full-landonline-data-dictionary-and-models.md
+	chmod +x $<
```
you can invoke `sh` directly here too, no need to `chmod +x`

Ashleigh234 replied:
If I don't use chmod +x I get `make: execvp: doc/tools/markdown-to-pdf-conversion.sh: Permission denied` problems.

strk replied:
Is that by doing: `bash doc/tools/markdown-to-pdf-conversion.sh ...` ?

strk commented on Makefile:
```diff
@@ -48,7 +57,38 @@ scripts/linz-lds-bde-schema-load: scripts/linz-lds-bde-schema-load.in Makefile
+		      doc/property-and-ownership-simplified-tables-data-dictionary.md
+	rm -f sql/comments.sql
+	python $< doc/lds-full-landonline-data-dictionary-and-models.md $@
+	python $< doc/property-and-ownership-simplified-tables-data-dictionary.md $@
```
Is the second command overriding the output of the first ?

Ashleigh234 replied:
The second command appends the additional comments to 08-lds_comments.sql, rather than overriding it.

strk replied:
Then a second call of that rule would keep appending data to the output, right ?
May I suggest having the python script write to standard output instead and use shell redirects from within the Makefile to make it clear what's going on ? In that case you would `>` in the first line and `>>` in the second one

strk commented on Makefile:
```diff
@@ -48,7 +57,38 @@ scripts/linz-lds-bde-schema-load: scripts/linz-lds-bde-schema-load.in Makefile
+	chmod +x $<
+	$< doc/property-and-ownership-simplified-tables-data-dictionary.md $@
+
+install: $(DOCS_built) $(SQLSCRIPTS) $(SCRIPTS_built)
```
Install rule does not really depend on $(DOCS_built) because it is not installing those docs ?
I suggest you add an `install-docs` rule, and make it a dependency of `install`

strk commented on doc/tools/markdown-to-commonmark-conversion.sh:
```diff
@@ -0,0 +1,31 @@
+# Perform pandoc markdown to commonmark conversion.
+pandoc temp.md -f \
+  markdown_strict+intraword_underscores+pipe_tables+hard_line_breaks+compact_definition_lists \
+  -t commonmark -o $2
```
You should check the return code from the pandoc call here, and `exit 1` if it was non-zero (HINT: `if test $? != 0`)

strk commented on doc/tools/markdown-to-commonmark-conversion.sh:
```diff
@@ -0,0 +1,31 @@
+  -t commonmark -o $2
+  
+# remove the temp file, no longer needed.
+rm temp.md
```
The cleanup would better go in a function invoked on exit (see `trap` bash keyword) - that way you can be sure to cleanup even if exiting prematurely (due to some error)

strk commented on doc/tools/markdown-to-commonmark-conversion.sh:
```diff
@@ -0,0 +1,31 @@
+# file suitable for read the docs.
+
+# Create temp markdown file
+cp $1 temp.md
```
better put temporary files under /tmp/, and ideally giving them a name which is unique to the process (you can find process id in the `$$` variable)

strk commented on doc/tools/markdown-to-pdf-conversion.sh:
```diff
@@ -0,0 +1,28 @@
+# Updates directory location of figures/ models to match correct location.
+
+# Create temp markdown file with current location. 
+cp $1 temp.md
```
Save file under /tmp/ and include PID in the filename

strk commented on doc/tools/markdown-to-pdf-conversion.sh:
```diff
@@ -0,0 +1,28 @@
+    markdown_strict+intraword_underscores+raw_tex+hard_line_breaks+pipe_tables+compact_definition_lists+yaml_metadata_block \
+    --toc --variable margin-left=0.75in --variable margin-right=0.75in         \
+    --variable margin-top=0.8in --variable margin-bottom=0.8in --toc-depth=2   \
+    -N --data-dir=""
```
Check return code from pandoc call and exit with a failure if it failed

strk commented on doc/tools/markdown-to-pdf-conversion.sh:
```diff
@@ -0,0 +1,28 @@
+    -N --data-dir=""
+
+# Remove the temp markdown file.
+rm temp.md
```
put cleanup in a function invoked on exit/error (see `trap`)

strk commented:
Can you think of a way to avoid using "wkhtmltopdf" ?
That'd be the preferred path, IMHO, as it keeps dependencies
low (note we also build debian packages so we want to only
rely on what's available as a package there, see debian/control)

Note that on an Ubuntu 17.10 (released on October 2017) the
available version for "pandoc" is 1.19.2. The one available
on Debian GNU/Linux 8.8 (latest stable) is 1.12.4.

Ideally we should require nothing older than 1.12.4

The script invoking "pandoc" should check for the version
of pandoc and give a meaningful error message to the user
when the version is not good enough.

Pandoc (including version) requirement should be documented
in the README.md file, under "Dependencies" section.

strk commented:
If there's NO WAY to do without `wkhtmltopdf` you could make
the PDF output optional (only generated when pandoc is recent
enough). In that case the "make" invocation should print a
clear WARNING about not building the PDF.

I see comments can be built w/out pandoc so I'm fine with that.

I'm intentionally not trying to figure out how to get a newer
`pandoc` installed on Travis because I think it is more
important for now to deal with the case in which pandoc is
not available or not recent enough.

Ashleigh234 commented:
There isn't really a way to not use 'wkhtmltopdf' and still get the pdf documents to style correctly. Even if 'wkhtmltopdf' could not be used, there are other changes from the earlier versions of pandoc that would produce other issues. I have worked out the earliest releases of pandoc that the pdf conversion works for (v 1.18)+ and the commonmark/ readthedocs conversion works for (v1.14)+. I have added these into the scripts so that if earlier versions of pandoc are used they produce warnings that they don't produce the PDF or commonmark markdown files. 

The Travis build is now failing in the same place I mentioned earlier, which happened after moving 07-lds-version.sql from the SQLSCRIPTS to SCRIPTS_built. With this error function lds.lds_version() does not exist.

strk commented:
Note I cleaned up some rules in the Makefile in master (thus the conflicts, I guess).
I'll be working in 12 hours from now while you'll probably be sleeping so if you want (let me know here) I can take what you have so far and rebase to master resolving conflicts and push a clean version of the work, then we see from there how to proceed.

Until then feel free to try the rebase/cleanup yourself and try making PDF generation optional as I was asking to do (with a big WARNING). Getting recent-enough pandoc to Travis can be done later.

strk commented:
Good catch about linz-lds-bde-schema-load missing scripts! The line should be extended to include `$(SQLSCRIPTS_built)`

Ashleigh234 commented:
I have fixed up the conflicts and updated the Makefile to add the `SQLSCRIPTS_built` into the linz-lds-bde-schema-load so the Travis build now works. It also produces the warnings if PDF generation isn't possible.

strk commented on doc/README.md:
```diff
@@ -0,0 +1,141 @@
+__Column Name__ | __Example Column 2__ | __Example Column 3__ | __Notes__
+:--------------|:-----------------|:------------|:-----------------------
+element_one    | field_2          | field_3     | element one notes
+element_two    | field_2a         | field_3a    | element two notes
```
I don't understand the "Example Column 2" and "Example Column 3" columns, what should go there ?
Example values ?

strk replied:
Also, how does the comment extractor script knows where a table description snippet starts and ends, in the markdown input ?

Ashleigh234 replied:
I have updated the document so that 'Example Column 2' and 'Example Column 3' aren't there anymore, rather is a short actual example.  

The comment extractor script checks for a second level header followed by a third level header with the name 'Description' to indicate the start of the table and checks for an empty line after entering the table to indicate the end of the table.
I updated it to include better instructions about where the table description starts and ends.

Ashleigh234 commented:
I changed the WARNING so that the generating pdf scripts exit with the error and if the error was caused by the pandoc version being too old, the makefile emits the warning.

strk commented:
Please make scripts more robust when called with wrong arguments. See what happens:
```
$ sh doc/tools/markdown-to-commonmark-conversion.sh
doc/tools/markdown-to-commonmark-conversion.sh: 14: [: 1: unexpected operator
cp: missing destination file operand after '/tmp/29822.md'
Try 'cp --help' for more information.
doc/tools/markdown-to-commonmark-conversion.sh: 21: doc/tools/markdown-to-commonmark-conversion.sh: cannot open /tmp/29822.md: No such file
pandoc: option `-o' requires an argument FILENAME
Try pandoc --help for more information.
```

A friendly script would print an Usage string on stderr instead (as I'm guessing it requires at least one parameter ?)

Same applies to all other scripts.

Another thing to be sure about is to exit on first failure, as it looks like multiple statements are being run (failing) before existing.

strk commented on doc/tools/markdown-to-commonmark-conversion.sh:
```diff
@@ -0,0 +1,48 @@
+
+if test $? != 0; then exit 1; fi
+  
+function finish(){
```
I think `function ` is not correct syntax (I get `Syntax error: "(" unexpected` when I try to use it). Should work with just `finish(){...}`

strk commented on doc/tools/markdown-to-commonmark-conversion.sh:
```diff
@@ -0,0 +1,48 @@
+function finish(){
+  rm /tmp/$$.md
+}
+trap finish EXIT
```
trap is too late in the file, will not catch any early exit, you should move all near the top of the file

strk commented on doc/tools/markdown-to-commonmark-conversion.sh:
```diff
@@ -0,0 +1,48 @@
+if test $? != 0; then exit 1; fi
+  
+function finish(){
+  rm /tmp/$$.md
```
This statement will fail if the finish function is called before `/tmp/$$.md` is created, I suggest adding a -f to the `rm` call. Also giving that file a prefix would help sysadmins understanding who's in charge of those temporary files :)

strk commented on doc/tools/markdown-to-commonmark-conversion.sh:
```diff
@@ -0,0 +1,48 @@
+fi
+fi
+
+if test $? != 0; then exit 1; fi
```
this check looks like being too late, why not checking right after `pandoc` call ?
Note that at that point you can also do:
`pandoc .... || exit 1`

strk commented on doc/tools/markdown-to-commonmark-conversion.sh:
```diff
@@ -0,0 +1,48 @@
+versionDecimal=$(echo $version | sed 's/\.//2g')
+versionRequired=1.14
+test=$(echo "$versionDecimal >= $versionRequired" | bc)
+if [ $test == 0 ]; then exit 5; else
```
You don't need an `else` after an `exit`. Omitting the else makes the code cleaner as you don't need to remember an additional `fi` later..

strk commented on doc/tools/markdown-to-commonmark-conversion.sh:
```diff
@@ -0,0 +1,48 @@
+version=$(pandoc --version | sed -n '1 p' | sed 's/.* //')
+versionDecimal=$(echo $version | sed 's/\.//2g')
+versionRequired=1.14
+test=$(echo "$versionDecimal >= $versionRequired" | bc)
```
The test looks fragile to me, how about just checking for major and minor when needed explictly ?
Full semver would be more complicated: https://github.com/cloudflare/semver_bash

strk commented on Makefile:
```diff
@@ -80,19 +147,11 @@ installcheck:
-	echo $$V && test "$$V" = "$(VERSION)" && \
-	V=`linz-lds-bde-schema-load --version` && \
-	echo $$V && test `echo "$$V" | awk '{print $$1}'` = "$(VERSION)"
+	psql -c 'select lds.lds_version()' linz-lds-bde-schema-test-db
```
(why) did you drop tests here ?

Ashleigh234 replied:
Didn't mean to drop the tests, I'll add them back in.

strk commented on Makefile:
```diff
@@ -80,19 +147,11 @@ installcheck:
-	echo $$V && test "$$V" = "$(VERSION)" && \
-	V=`linz-lds-bde-schema-load --version` && \
-	echo $$V && test `echo "$$V" | awk '{print $$1}'` = "$(VERSION)"
+	psql -c 'select lds.lds_version()' linz-lds-bde-schema-test-db
```
(why) did you drop tests here ?

strk commented on .travis.yml:
```diff
@@ -49,7 +49,7 @@ matrix:
   # Install common packages
   - sudo apt-get -qq update
-  - sudo apt-get install -y pgtap pgxnclient
+  - sudo apt-get install -y pgtap pgxnclient pandoc
```
this, we found, is not good enough for Travis to get a working pandoc, time to find out how to install one ?

strk commented on Makefile:
```diff
@@ -50,23 +58,82 @@ scripts/linz-lds-bde-schema-load: scripts/linz-lds-bde-schema-load.in Makefile
+install-docs: $(DOCS_built)
+	mkdir -p ${datadir}/docs
+	find \doc -name \*rtd.md -exec cp {} ${datadir}/docs \;
+	find \doc -name \*.pdf -exec cp {} ${datadir}/docs \;
```
why not using $(DOCS_built) variable instead of `find` ? The former would be more strict

Ashleigh234 replied:
I tried using the $(DOCS_built) variable but produced errors if the PDF or markdown files didn't exist, because they weren't created due to the pandoc version.

strk commented:
> I tried using the $(DOCS_built) variable but produced errors if the PDF or markdown files didn't exist, because they weren't created due to the pandoc version.

Makefile supports conditionals, so you could have a variable
express whether or not "pandoc" dependency is met and use it
to enable/disable/replace-with-a-warning whole rules.

strk commented:
It looks like wkhtmltopdf is not part of `pandoc` installation, so you should probably mention it as a dependency in docs. Debian 8 packages version 0.12.1 of it, it is hopefully enough ?

imincik commented:
@Ashleigh234 , @strk we need to be able to build packages for Ubuntu 14.04 which is still our prod OS and is containing Pandoc 1.12.

Ashleigh234 commented:
I have updated the dependency in the readme. I haven't had a problem running wkhtmltopdf locally, but have found that it relies on xserver to be installed. Unfortunately only using pandoc by it self doesn't generate the pdf's, it requires either latex or wkhtmltopdf. Wkhtmltopdf was being used because it was the only way to format tables in the pdf, by using a stylesheet.

strk commented:
So the plan is:

 - have travis install new enough pandoc 
 - have PDF built from a separate rule (not "make") so that it is only built on demand
 - do not include the PDF in the core debian package (a separate package can be built when pandoc is available)

strk commented on Makefile:
```diff
@@ -15,42 +15,66 @@ bindir=${DESTDIR}/usr/bin
 
 scripts/linz-lds-bde-schema-load: scripts/linz-lds-bde-schema-load.in Makefile
-	$(SED) -e 's|@@SQLSCRIPTS@@|$(SQLSCRIPTS) $(SQLSCRIPTS_built)|' \
+	$(SED) -e 's|@@SQLSCRIPTS@@|$(SQLSCRIPTS)$(SQLSCRIPTS_built)|' \
```
I think you do want that space between `$(SQLSCRIPTS)` and `$(SQLSCRIPTS_built)`, lack of space relies on how you define those variables (if with a trailing or leading space or not)

strk commented on Makefile:
```diff
@@ -61,7 +85,7 @@ install: $(SQLSCRIPTS) $(SQLSCRIPTS_built) $(SCRIPTS_built)
 	rm -rf ${datadir}
 
-check test: $(SQLSCRIPTS) $(SQLSCRIPTS_built)
+check test: $(SQLSCRIPTS) check-docs
```
the "check" rule should only check what was built with the default rule, so if docs are not built by simply running "make" then "check" should not be checking the docs (how is it possible that travis passes `check-docs` if it is not building them ?)

strk replied:
Ah sorry, now I realized that rule is checking the sources, not the built docs

strk commented on Makefile:
```diff
@@ -96,3 +120,56 @@ installcheck:
 	echo $$V && test `echo "$$V" | awk '{print $$1}'` = "$(VERSION)"
 	dropdb linz-lds-bde-schema-test-db
+
+create-docs: $(DOCS_built)
```
I would rename this target to `docs` (for `make docs`) - or add an alias, like for the `check test` target

strk commented on README.md:
```diff
@@ -59,8 +59,10 @@ Dependencies
+[dbpatch](https://github.com/linz/postgresql-dbpatch) (v1.2.0+),
+[linz_bde_uploader](https://github.com/linz/linz_bde_uploader) (v2.4.0+),
+[pandoc](https://github.com/jgm/pandoc) (v1.18+) and
+[wkhtmltopdf](https://github.com/wkhtmltopdf/wkhtmltopdf) (v0.12.1+)
```
Please document that pandoc and wkhtmltopdf are only required to build documentation, so they are not "strong" dependencies

strk commented:
reguarding those PNG: if they were SVG I would feel better, would the PDF generation rule support that ?

strk commented:
Ok I've done all of the things I were requesting myself. In current status the branch has Travis checking building w/out docs dependencies and then building the docs after installing deps. README was updated to be more informative about dependencies and docs building.

All is left to do, beside the PNG -> SVG idea, is to interactively rebase to master cleaning up some noise (like addition/removal of files)

strk commented:
I'm dealing with the branch cleanup now, will squash all your initial work in a single commit and then append my commits on top.

imincik commented:
Thanks @Ashleigh234 and @strk
